### PR TITLE
Refactoring to work with PHENIO and use more generalizable frequencies

### PR DESCRIPTION
**Pull request by caufieldjh**
As per plan suggested by @justaddcoffee:

- [x] ingest Phenio
- [x] preprocess it into a DAG
      Filter to subclass_of edges, transpose, and remove disconnected nodes
- [x]  calculate frequencies assuming 1 occurrence per class
- [x] do an all by all comparison of all phenotypes in phenio
- [x] write each phenotype pair to file IFF the pair is greater than some cutoff (2.5 Resnik similarity?)
- [x] add validation checks re: assumptions here https://github.com/Knowledge-Graph-Hub/semsim/pull/4#issuecomment-1271968027

### CODE REVIEW

**Comment by justaddcoffee** [comment]
Prototype colab notebook for this is [here](https://colab.research.google.com/drive/1m8sQM5qf8-C3FsVOcYW4j_Ji28mMi7_V?usp=sharing)

@caufieldjh @LucaCappelletti94

**Comment by caufieldjh** [comment]
Jaccard completes as expected, though Resnik is still presenting difficulties.
Initially this was because the input graph contained two components after filtering (the smaller one being just 13 nodes, so it seems fine to remove it).
Using just the largest connected component, we still see errors about lack of shared root nodes, e.g., for [MP:0003708](https://www.ebi.ac.uk/ols/ontologies/mp/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FMP_0003708) and HP:0000001 (the HP root). There are multiple roots, like the MP root, MP:0000001, but that's still a subclass_of UPHENO:0001003, itself a subclass of UPHENO:0001001, itself a subclass of PATO:0000001. The HP root is a subclass_of owl:Thing. 
So the take-home here is that there should be at least one common parent between all HP and MP terms, but the graph does have multiple roots, at least if we don't include a full path for everything back to owl:Thing.

The solution is to include PATO, owl, and BFO along with HP, MP, and UPHENO.

**Comment by LucaCappelletti94** [comment]
Hello Harry, do you need a hand with this work or is all proceeding nominally? Maybe we should define a better behaviour than "crash" for the cases without a common root?

**Comment by caufieldjh** [comment]
Thanks, @LucaCappelletti94 - it looks like it's working correctly for now. 
I suspect that filtering based on prefix will often lead to cases like this, in which we don't actually know how the root is defined and we're mostly interested in semantic similarity downstream from that root. For most ontologies, this shouldn't be an issue, but any KG or integrated set of ontologies *will* have this issue. 

Throwing a ValueError may still be the best option, but an alternative would be to allow `get_pairwise_similarities()` to take a prefix filter argument and expect the input DAG to be the un-prefix-filtered graph. Then if there are multiple roots encountered it can suggest expanding the set of nodes to cover. 

I wouldn't want to cover the whole graph by default, though: just the subset of KG-Phenio used above yields a Resnik score set of >24G and about 2.6 billion values in all.

**Comment by caufieldjh** [comment]
The downside to iterating instead of working on the whole graph (or filtered subgraph) is that it requires much more memory and is slower.

**Comment by justaddcoffee** [comment]
- TODO: via @cmungall: include these edges too when making DAG: `part_of`, `is_phenotypic_??` UPHENO_0001/3?

**Comment by caufieldjh** [comment]
Just going to treat all Resnik values as numpy arrays - somewhere around 50K pairs this becomes a memory issue otherwise

**Comment by cmungall** [comment]
Would it help to filter further upstream in the rust code?

**Comment by caufieldjh** [comment]
> Would it help to filter further upstream in the rust code?
Yes, and Luca has plans to add it in the call to `get_pairwise_similarities` - having a smaller DF would make the larger operations tractable since the bottleneck is running operations over the DF, but pandas really wants to load it all. Probably makes sense to do more chunking prior to producing output, too

**Comment by LucaCappelletti94** [comment]
@cmungall and @caufieldjh, I have added the parameters to filter Resnik by minimum threshold values and skipping values with undefined Resnik scores (those with different DAG roots). As soon as I finish testing it, I will push it to PyPI and update the code snippets I wrote that use those APIs in this repo.

**Comment by justaddcoffee** [comment]
Thanks @LucaCappelletti94!

**Comment by caufieldjh** [comment]
Thanks, @LucaCappelletti94 !

**Comment by justaddcoffee** [comment]
@hrshdhgd when converting that sparse pd dataframe to a "long" dataframe, possibly [`stack`](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.stack.html) might be faster than `melt` in this case (although I'm not sure - haven't compared them)

**Comment by hrshdhgd** [comment]
I've added code to both. We can decide which one to use when we deal with real-world data.

**Comment by caufieldjh** [comment]
Still runs out of memory on KG-Phenio before writing the Resnik output and almost immediately after completing the `get_pairwise_similarities` operation.

**Comment by caufieldjh** [comment]
Bit of an odd thing I just saw: if I set the `PREFIXES` constant (i.e., the list of prefixes to traverse but not collect scores for) to an empty list and run `semsim sim KGPhenio -p HP,MP` then everything completes as expected with no errors but only includes HP vs. HP scores. My thought is that the initial DAG gets over-filtered, so the largest component (~16K nodes) doesn't contain a single MP node.

If I include UPHENO in the `PREFIXES` then the input DAG is much larger - 49291 nodes in the largest component. This throws the "no shared parent node" error because it's still missing BFO etc.

So it's probably not ideal to *always* include other prefixes if we don't really need them to traverse from one node type vs. another, but if we do need them then we don't currently have any way of knowing that as no warnings are raised.
One way to address this may be to list the prefixes present in each component and throw a warning/error if we throw out a component containing nodes we need scores for.

**Comment by caufieldjh** [comment]
That last commit definitely won't work with current `embiggen`, but it should work once that's available.

**Comment by caufieldjh** [comment]
Also, transform of ero fails silently

**Comment by caufieldjh** [comment]
This is much faster for single ontologies (e.g., HP) but KGPhenio still doesn't complete on my local machine, so I tried it on a 32 GB GCP instance. It looks like it took about 1 hr 45 min but then failed at some point after calculating Resnik - either joining the Jaccard values or while writing the output. Manageable issues either way.

**Comment by caufieldjh** [comment]
Tried it again on GCP - this time it took ~5 hrs and then died before starting the Jaccard computation 🤨

File: semsim/compute_pairwise_similarities.py [hunk 1]
```diff
@@ -69,9 +67,15 @@ def compute_pairwise_sims(
...
+        # Remap node IDs to node names
+        node_map = dag.get_nodes_mapping()
+        id_map = {v: k for k, v in node_map.items()}
+        for col in ['source', 'destination']:
+            rs_df = rs_df.replace({col: id_map})
+
         print(f"Writing output to {rs_path}...")
         rs_df.sort_values(
-            by=["similarity"], ascending=False, inplace=True
+            by=["resnik_score"], ascending=False, inplace=True
```
Comment (matentzn) [inline semsim/compute_pairwise_similarities.py:78]: I am still very against this design... :P Not your fault, need to hash this out with @cmungall. Having the a generic `score` column is much more flexible than encoding all scores in different columns

Comment (caufieldjh) [reply semsim/compute_pairwise_similarities.py:75]: I like to know what kind of score it is. Could we include a `score_types` column or something?

Comment (matentzn) [reply semsim/compute_pairwise_similarities.py:75]: In OAK you dont have a choice. Whatever the semsim model says... :/

**Comment by caufieldjh** [comment]
This PR checks all the boxes now...as long as the cutoff isn't as low as 2.5. I used a cutoff of 5 and got KGPhenio HP vs MP results in a couple minutes (!!!).
They are here, for reference: https://kg-hub.berkeleybop.io/kg-phenio/hp-mp_similarities.tar.gz

**Comment by caufieldjh** [comment]
Resnik scores make sense, but need to verify that Jaccard scores are getting calculated as expected.
Many high-scoring pairs have Jaccard of zero.
See `MP:0002904,HP:0003165` for example - Resnik of 13.833021 but Jaccard of zero.

**Comment by cmungall** [comment]
seems that all inter-ontology links have jaccard 0

**Comment by caufieldjh** [comment]
@LucaCappelletti94 suggested that the Jaccard zeros may be related to the choice of root node. We try to enforce using a single root but that means `HP:0000001` may get selected; we'd prefer to use something HP and MP have in common, like the UPHENO root. I've confirmed that this solves the issue, as using that root gets me:
```
MP:0002904,HP:0003165,13.76679,0.3
```

**Comment by cmungall** [comment]
avoid making any assumptions about root nodes.

Some false assumptions about ontologies:

- there is a single root node (unless owl:Thing is considered)
- there is a single graph
- the graph of all predicate types is a DAG
- the is-a graph is a DAG (subClassOf is reflexive, A=B <=> A sub B, B sub A
etc

**Comment by caufieldjh** [comment]
Try repeating the Jaccard operation if there is >1 root

**Comment by caufieldjh** [comment]
Handling assumptions:
* Multiple nodes are OK - we compute over one or >1.
* Single graph - we take the largest connected component right now. We may have other non-trivially large connected components, but should we compute over them on their own? We should discard singletons, at least.
* If graph isn't a DAG we encounter trouble in calculating similarity.
* If subgraphs based on predicate type aren't DAGs, same thing, but we may need more options to filter (a la OAK and ROBOT)

**Comment by caufieldjh** [comment]
Continuing discussion of assumptions re: semsim here: https://github.com/Knowledge-Graph-Hub/semsim/issues/11

**Pull request merged**